feat(core): expose new input API for signal-based inputs

devversion opened the pull request:
Enables signal inputs for existing Zone based components.
This is a next step we are taking to bring signal inputs earlier to the Angular community.

The goal is to enable early access for the ecosystem to signal inputs, while we are continuing
development of full signal components as outlined in the RFC. This will allow the ecosystem
to start integrating signals more deeply, prepare for future migrations, and improves code quality
and DX for existing components (especially for OnPush).

Based on our work on full signal components, we've gathered more information and learned
new things. We've improved the API by introducing a way to intuitively declare required inputs,
as well as improved the API around initial values. We even support non-primitive initial values
as the first argument to the `input` function now.

```ts
@Directive({..})
export class MyDir {
  firstName = input<string>();            // string|undefined
  lastName = input.required<string>();    // string
  age = input(0);                         // number
```

devversion committed:
refactor(compiler-cli): detect input functions without partial evaluation

This allows us to ensure signal inputs and a potential JIT transform
remain single file compilation compatible. The consequences are that
options need to be statically analyzable more strictly, compared to
loosened restrictions with static interpretation where e.g. `alias`
can be defined through a shared variable.

devversion committed:
feat(core): expose new `input` API for signal-based inputs

Enables signal inputs for existing Zone based components.
This is a next step we are taking to bring signal inputs earlier to the Angular community.

The goal is to enable early access for the ecosystem to signal inputs, while we are continuing
development of full signal components as outlined in the RFC. This will allow the ecosystem
to start integrating signals more deeply, prepare for future migrations, and improves code quality
and DX for existing components (especially for OnPush).

Based on our work on full signal components, we've gathered more information and learned
new things. We've improved the API by introducing a way to intuitively declare required inputs,
as well as improved the API around initial values. We even support non-primitive initial values
as the first argument to the `input` function now.

```ts
@Directive({..})
export class MyDir {
  firstName = input<string>();            // string|undefined
  lastName = input.required<string>();    // string
  age = input(0);                         // number
```

devversion commented:
FYI: Removing CLI shared review-request as this is a non-CLI facing change in `tooling.ts`, and we have a pre-existing review on signal inputs in a separate PR

alxhub reviewed (approved):
Reviewed-for: public-api

devversion committed:
fixup! feat(core): expose new `input` API for signal-based inputs

Remove unnecessary exports for API guarding

Note: Need to update dev-infra to avoid a API extractor issue.

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/input_function.ts:
```diff
@@ -108,8 +107,32 @@ function isReferenceToInputFunction(
+  if (!ts.isObjectLiteralExpression(optionsNode)) {
+    throw new FatalDiagnosticError(
+        ErrorCode.VALUE_HAS_WRONG_TYPE, optionsNode,
+        'Argument needs to be an object literal that is statically analyzable.');
```
nit (not blocking): I'd remove the "statically analyzable" part here, that isn't telling users much compared to the "needs to be an object literal"

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

atscott commented:
This PR was merged into the repository by commit 863be4b6981dc60ca0610b0e61d2ba1f5759e2a3.

maximLyakhov commented:
So we can't use `.set` on these inputs?

eneajaho commented:
> So we can't use `.set` on these inputs?

Jo. There will be a new kind of input called model input that will serve for two way data binding (and that you can call set on it)

iamart commented:
What trips me a bit is that `input()` takes raw value and not a "Signal" being passed to it; that is, you bind it like `[prop]="signal()"` and not `[prop]="signal"`. Feels like binding a value directly but hopefully change detection just handles well.
